src/libraries/System.Security.Cryptography/tests/OpenSslNamedKeysTests.manual.cs

```diff
@@ -174,17 +225,202 @@ public static void Engine_UsePublicKey()
+            }
+
+            using (ECDiffieHellmanPublicKey alicePublic = alicePub.PublicKey)
+            {
+                byte[] sharedKeyFromBob = bobPri.DeriveKeyMaterial(alicePublic);
+                Assert.Equal(sharedKeyFromAlice, sharedKeyFromBob);
+            }
+        }
+
+        [ConditionalFact(nameof(ShouldRunProviderRsaSignTests))]
+        [ActiveIssue("https://github.com/tpm2-software/tpm2-openssl/issues/115")]
+        // or a workaround API proposal
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/104080")]
+        public static void Provider_TPM2SignRsa()
+        {
+            using SafeEvpPKeyHandle priKeyHandle = SafeEvpPKeyHandle.OpenKeyFromProvider(Tpm2ProviderName, TpmRsaSignKeyHandleUri);
+            using RSA rsaPri = new RSAOpenSsl(priKeyHandle);
+            byte[] rsaPubBytes = rsaPri.ExportSubjectPublicKeyInfo();
+            RSA rsaPub = RSA.Create();
+            rsaPub.ImportSubjectPublicKeyInfo(rsaPubBytes, out int bytesRead);
+            Assert.Equal(rsaPubBytes.Length, bytesRead);
+
+            using RSA rsaBad = RSA.Create();
+            rsaBad.KeySize = rsaPri.KeySize;
+
+            byte[] data = new byte[] { 1, 2, 3, 1, 1, 2, 3 };
+            byte[] badSignature = rsaBad.SignData(data, HashAlgorithmName.SHA256, RSASignaturePadding.Pss);
+
+            // can use same key more than once
+            for (int i = 0; i < 10; i++)
+            {
+                data[0] = (byte)i;
+                byte[] signature = rsaPri.SignData(data, HashAlgorithmName.SHA256, RSASignaturePadding.Pss);
+                Assert.True(rsaPub.VerifyData(data, signature, HashAlgorithmName.SHA256, RSASignaturePadding.Pss), "signature does not verify with the right key");
+                Assert.False(rsaPub.VerifyData(data, badSignature, HashAlgorithmName.SHA256, RSASignaturePadding.Pss), "signature should not verify with the wrong key");
+
+                signature[12] ^= 1;
+                Assert.False(rsaPub.VerifyData(data, signature, HashAlgorithmName.SHA256, RSASignaturePadding.Pss), "tampered signature should not verify");
+
+                // TPM key is intended for sign only, we could theoretically make verify work without needing to export/import by forcing 'default' provider
+                // for this operation it's most likely misusage on user part and tpm2 provider intentionally didn't allow it so we will follow this logic.
+                Assert.ThrowsAny<CryptographicException>(() => rsaPri.VerifyData(data, signature, HashAlgorithmName.SHA256, RSASignaturePadding.Pss));
+            }
+        }
+
+        [ConditionalFact(nameof(ShouldRunProviderRsaDecryptTests))]
+        public static void Provider_TPM2DecryptRsa()
+        {
+            // TPM2 OAEP support was added in the second half of 2023 therefore we only test Pkcs1 padding
+            // See: https://github.com/tpm2-software/tpm2-openssl/issues/89
```

Comment:
That's practically ancient history!

In all seriousness, an OAEP test could be written to show that it either works as expected, or throws on decrypt, but that we're never mishandling the data and giving bad results.

Reply:
done, now test will pass when decrypt fails with OAEP or return correct result